python/cudf/cudf/core/frame.py

```diff
@@ -3337,18 +3329,240 @@ def _reindex(
+                    right_column, dtype=left_column.dtype
+                )
+            elif not isinstance(right_column, ColumnBase):
+                right_column = left_column.normalize_binop_value(right_column)
+
+            fn_apply = fn
+            if fn == "truediv":
+                # Decimals in libcudf don't support truediv, see
+                # https://github.com/rapidsai/cudf/pull/7435 for explanation.
+                if is_decimal_dtype(left_column.dtype):
+                    fn_apply = "div"
+
+                # Division with integer types results in a suitable float.
+                truediv_type = {
+                    np.int8: np.float32,
+                    np.int16: np.float32,
+                    np.int32: np.float32,
+                    np.int64: np.float64,
+                    np.uint8: np.float32,
+                    np.uint16: np.float32,
+                    np.uint32: np.float64,
+                    np.uint64: np.float64,
+                    np.bool_: np.float32,
+                }.get(left_column.dtype.type)
+                if truediv_type is not None:
+                    left_column = left_column.astype(truediv_type)
+
+            output_mask = None
+            if fill_value is not None:
+                if is_scalar(right_column):
+                    if left_column.nullable:
+                        left_column = left_column.fillna(fill_value)
+                else:
+                    # If both columns are nullable, pandas semantics dictate
+                    # that nulls that are present in both left_column and
+                    # right_column are not filled.
+                    if left_column.nullable and right_column.nullable:
+                        lmask = as_column(left_column.nullmask)
+                        rmask = as_column(right_column.nullmask)
+                        output_mask = (lmask | rmask).data
+                        left_column = left_column.fillna(fill_value)
+                        right_column = right_column.fillna(fill_value)
+                    elif left_column.nullable:
+                        left_column = left_column.fillna(fill_value)
+                    elif right_column.nullable:
+                        right_column = right_column.fillna(fill_value)
+
+            # For bitwise operations we must verify whether the input column
+            # types are valid, and if so, whether we need to coerce the output
+            # columns to booleans.
+            coerce_to_bool = False
+            if fn_apply in {"and", "or", "xor"}:
+                err_msg = (
+                    f"Operation 'bitwise {fn_apply}' not supported between "
+                    f"{left_column.dtype.type.__name__} and {{}}"
+                )
+                if right_column is None:
+                    raise TypeError(err_msg.format(type(None)))
+
+                try:
+                    left_is_bool = np.issubdtype(left_column.dtype, np.bool_)
```

Comment:
I think we should move away from any new uses of numpy dtype APIs in favor of creating/using our own internal utility, even if that function still uses the numpy function for now.

Reply:
Yeah this is on me, I got burnt out on the pandas/numpy type API refactor after writing that one massive one with all the tests for pandas.api.types. I'll do this and remember to continue doing this incrementally in the future.